File: samples/app-complete-auth/csharp/AppCompleteAuth/Dialogs/MainDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using AppCompleteAuth.Models;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Configuration;

namespace AppCompleteAuth.Dialogs
{
    /// <summary>
    /// MainDialog class to handle the main dialog flow.
    /// </summary>
    public class MainDialog : ComponentDialog
    {
        private readonly string _applicationBaseUrl;
        private readonly ConcurrentDictionary<string, Token> _token;

        public MainDialog(IConfiguration configuration, ConcurrentDictionary<string, Token> token)
            : base(nameof(MainDialog))
        {
            _token = token;
            _applicationBaseUrl = configuration["ApplicationBaseUrl"] ?? throw new NullReferenceException("ApplicationBaseUrl");

            AddDialog(new TextPrompt(nameof(TextPrompt)));
            AddDialog(new FacebookAuthDialog(configuration["FacebookConnectionName"], _token));
            AddDialog(new BotSsoAuthDialog(configuration["ConnectionName"], _token));
            AddDialog(new UsernamePasswordAuthDialog(configuration["ApplicationBaseUrl"]));
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
            {
                    PromptStepAsync
            }));

            // The initial child Dialog to run.
            InitialDialogId = nameof(WaterfallDialog);
        }

        /// <summary>
        /// Method to invoke auth flow.
        /// </summary>
        private async Task<DialogTurnResult> PromptStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var text = stepContext.Context.Activity.Text?.ToLower().Trim();

            if (text == null || text == "usingcredentials")
            {
                return await stepContext.BeginDialogAsync(nameof(UsernamePasswordAuthDialog));
            }
            if (text == "sso" || text == "logoutsso")
            {
                return await stepContext.BeginDialogAsync(nameof(BotSsoAuthDialog));
            }
            if (text == "facebooklogin" || text == "logoutfacebook")
            {
                return await stepContext.BeginDialogAsync(nameof(FacebookAuthDialog));
            }

            await stepContext.Context.SendActivityAsync(MessageFactory.Attachment(GetLoginOptionCard()), cancellationToken);
            return await stepContext.EndDialogAsync();
        }

        /// <summary>
        /// Get login option card.
        /// </summary>
        private static Attachment GetLoginOptionCard()
        {
            var heroCard = new HeroCard
            {
                Title = "Login options",
                Text = "Select a login option",
                Buttons = new List<CardAction>
                    {
                        new CardAction(ActionTypes.MessageBack, title: "AAD SSO authentication", value: "sso", text: "sso", displayText: "AAD SSO authentication"),
                        new CardAction(ActionTypes.MessageBack, title: "Facebook login (OAuth 2)", value: "facebooklogin", text: "facebooklogin", displayText: "Facebook login (OAuth 2)"),
                        new CardAction(ActionTypes.MessageBack, title: "User Id/password login", value: "usingcredentials", text: "usingcredentials", displayText: "User Id/password login"),
                    }
            };

            return heroCard.ToAttachment();
        }

        /// <summary>
        /// Get sign in card.
        /// </summary>
        private Attachment GetPopUpSignInCard()
        {
            var heroCard = new HeroCard
            {
                Title = "Sign in card",
                Buttons = new List<CardAction>
                    {
                        new CardAction(ActionTypes.Signin, "Sign in", value: $"{_applicationBaseUrl}/popUpSignin?from=bot&height=535&width=600"),
                    }
            };

            return heroCard.ToAttachment();
        }
    }
}

File: samples/app-complete-auth/nodejs/dialogs/mainDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { DialogSet, DialogTurnStatus, WaterfallDialog, ComponentDialog } = require('botbuilder-dialogs');
const { CardFactory, ActionTypes } = require('botbuilder');

const MAIN_DIALOG = 'MainDialog';
const MAIN_WATERFALL_DIALOG = 'MainWaterfallDialog';

const { SimpleFacebookAuthDialog } = require('./simpleFacebookAuthDialog');
const { UserNamePasswordDialog } = require('./userNamePasswordDialog');
const { BotSSOAuthDialog } = require('./botSSOAuthDialog');

const FACEBOOKAUTH = 'FacebookAuth';
const USERNAMEPASSWORD = 'UserNamePassword';
const SSOAUTH = 'SsoAuth';

/**
 * MainDialog class handles the main dialog process.
 */
class MainDialog extends ComponentDialog {
    constructor() {
        super(MAIN_DIALOG, process.env.connectionName);
        this.baseUrl = process.env.ApplicationBaseUrl;

        this.addDialog(new SimpleFacebookAuthDialog(FACEBOOKAUTH));
        this.addDialog(new BotSSOAuthDialog(SSOAUTH));
        this.addDialog(new UserNamePasswordDialog(USERNAMEPASSWORD));
        this.addDialog(new WaterfallDialog(MAIN_WATERFALL_DIALOG, [
            this.promptStep.bind(this)
        ]));

        this.initialDialogId = MAIN_WATERFALL_DIALOG;
    }

    /**
     * The run method handles the incoming activity (in the form of a DialogContext) and passes it through the dialog system.
     * If no dialog is active, it will start the default dialog.
     * @param {Object} context - The context of the current turn of conversation.
     * @param {Object} accessor - The state accessor for the dialog.
     */
    async run(context, accessor) {
        const dialogSet = new DialogSet(accessor);
        dialogSet.add(this);
        const dialogContext = await dialogSet.createContext(context);
        const results = await dialogContext.continueDialog();

        if (results.status === DialogTurnStatus.empty) {
            await dialogContext.beginDialog(this.id);
        }
    }

    /**
     * The promptStep method handles the initial prompt step of the dialog.
     * @param {Object} stepContext - The context of the current step.
     */
    async promptStep(stepContext) {
        try {
            const text = stepContext.context._activity.text;
            if (!text || text.trim() === "usingcredentials") {
                return await stepContext.beginDialog(USERNAMEPASSWORD);
            } else if (text.trim() === "sso") {
                return await stepContext.beginDialog(SSOAUTH);
            } else if (text.trim() === "facebooklogin") {
                return await stepContext.beginDialog(FACEBOOKAUTH);
            } else {
                const buttons = [
                    { type: ActionTypes.ImBack, title: 'AAD SSO authentication', value: 'sso' },
                    { type: ActionTypes.ImBack, title: 'Facebook login (OAuth 2)', value: 'facebooklogin' },
                    { type: ActionTypes.ImBack, title: 'User Id/password login', value: 'usingcredentials' }
                ];
                const card = CardFactory.heroCard('Login options', undefined, buttons, { text: "Select a login option" });
                await stepContext.context.sendActivity({ attachments: [card] });
                return await stepContext.endDialog();
            }
        } catch (err) {
            console.error(err);
        }
    }
}

module.exports.MainDialog = MainDialog;